File: neetcode/go/_347/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(topKFrequent([]int{1, 1, 1, 2, 2, 3}, 2))
}

func topKFrequent(nums []int, k int) []int {
	mp := make(map[int]int)

	for _, v := range nums {
		mp[v]++
	}

	frequent := make([][]int, len(nums)+1)
	for k, v := range mp {
		frequent[v] = append(frequent[v], k)
	}

	result := make([]int, 0, k)
	for i := len(nums); i >= 0; i-- {
		if len(frequent[i]) == 0 {
			continue
		}

		for _, v := range frequent[i] {
			result = append(result, v)
			if len(result) == k {
				return result
			}
		}
	}

	return []int{}
}

File: neetcode/python/_347/main.py
from heapq import heappush, heappop
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        mp = dict()
        for num in nums:
            if num in mp:
                mp[num] += 1
            else:
                mp[num] = 1
        
        heap = []
        for key, c in mp.items():
            heappush(heap, (-c, key))
        
        res = []
        for i in range(k):
            v = heappop(heap)
            res.append(v[1])
        return res

print(Solution().topKFrequent([1, 1, 1, 2, 2, 3], 2))